File: libs/server/persistent/src/persistent/prisma/workspace-members.ts
import { randomUUID } from 'node:crypto';
import {
  DomainError,
  Member,
  MemberDescription,
  WorkspaceMembers,
} from '@evidence/server-domain';
import { EntityList } from '../database';
import { assembleMember } from './mappers';
import type { PrismaStore } from './types';
import { defaultIfBlank, isUniqueConflict, now } from './utils';

export class PrismaWorkspaceMembers
  extends EntityList<Member>
  implements WorkspaceMembers
{
  constructor(
    private readonly store: PrismaStore,
    private readonly workspaceId: string,
  ) {
    super();
  }

  protected override async findEntities(
    from: number,
    to: number,
  ): Promise<Member[]> {
    const rows = await this.store.workspaceMember.findMany({
      where: { workspaceId: this.workspaceId },
      orderBy: { createdAt: 'asc' },
      skip: from,
      take: Math.max(to - from, 0),
    });
    return rows.map(assembleMember);
  }

  protected override async findEntity(id: string): Promise<Member | null> {
    const row = await this.store.workspaceMember.findFirst({
      where: { id, workspaceId: this.workspaceId },
    });
    return row ? assembleMember(row) : null;
  }

  override async size(): Promise<number> {
    return this.store.workspaceMember.count({
      where: { workspaceId: this.workspaceId },
    });
  }

  async addMember(desc: MemberDescription): Promise<Member> {
    const workspaceId = desc.workspace.id();
    if (workspaceId !== this.workspaceId) {
      throw DomainError.validation(
        `member workspace ${workspaceId} does not match scoped workspace ${this.workspaceId}`,
      );
    }

    const userId = desc.user.id();
    const [user, workspace] = await Promise.all([
      this.store.user.findUnique({ where: { id: userId } }),
      this.store.workspace.findFirst({
        where: { id: this.workspaceId, deletedAt: null },
      }),
    ]);
    if (!user) {
      throw DomainError.notFound(`user ${userId} not found`);
    }
    if (!workspace) {
      throw DomainError.notFound(`workspace ${this.workspaceId} not found`);
    }

    const timestamp = now();
    try {
      const row = await this.store.workspaceMember.create({
        data: {
          id: randomUUID(),
          workspaceId: this.workspaceId,
          userId,
          role: defaultIfBlank(desc.role, 'member'),
          createdAt: timestamp,
          updatedAt: timestamp,
        },
      });
      return assembleMember(row);
    } catch (error) {
      if (isUniqueConflict(error)) {
        throw DomainError.conflict(
          `user ${userId} is already a workspace member`,
        );
      }
      throw error;
    }
  }

  async updateMember(memberId: string, role: string): Promise<Member> {
    const row = await this.requireMember(memberId);
    const normalizedRole = defaultIfBlank(role, '');
    if (!normalizedRole) {
      throw DomainError.validation('workspace member role must not be empty');
    }
    await this.assertOwnerRemains(row.role, normalizedRole);
    const updated = await this.store.workspaceMember.update({
      where: { id: row.id },
      data: { role: normalizedRole, updatedAt: now() },
    });
    return assembleMember(updated);
  }

  async removeMember(memberId: string): Promise<void> {
    const row = await this.requireMember(memberId);
    await this.assertOwnerRemains(row.role, null);
    await this.store.workspaceMember.delete({ where: { id: row.id } });
  }

  private async requireMember(memberId: string) {
    const row = await this.store.workspaceMember.findFirst({
      where: { id: memberId, workspaceId: this.workspaceId },
    });
    if (!row) {
      throw DomainError.notFound(`workspace member ${memberId} not found`);
    }
    return row;
  }

  private async assertOwnerRemains(
    currentRole: string,
    nextRole: string | null,
  ): Promise<void> {
    if (currentRole !== 'owner' || nextRole === 'owner') {
      return;
    }
    const owners = await this.store.workspaceMember.count({
      where: { workspaceId: this.workspaceId, role: 'owner' },
    });
    if (owners <= 1) {
      throw DomainError.conflict('workspace must retain at least one owner');
    }
  }
}

File: libs/server-java/persistent/src/main/java/reengineering/ddd/evidence/persistent/associations/WorkspaceMembers.java
package reengineering.ddd.evidence.persistent.associations;

import io.github.jayclock.smartdomain.mybatis.AssociationMapping;
import io.github.jayclock.smartdomain.mybatis.database.EntityList;
import jakarta.inject.Inject;
import java.time.Clock;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.UUID;
import org.springframework.dao.DuplicateKeyException;
import reengineering.ddd.evidence.domain.DomainException;
import reengineering.ddd.evidence.domain.description.MemberDescription;
import reengineering.ddd.evidence.domain.model.Member;
import reengineering.ddd.evidence.domain.model.Workspace;
import reengineering.ddd.evidence.domain.validation.WorkspaceAccess;
import reengineering.ddd.evidence.persistent.mappers.UsersMapper;
import reengineering.ddd.evidence.persistent.mappers.WorkspaceMembersMapper;
import reengineering.ddd.evidence.persistent.mappers.WorkspacesMapper;

@AssociationMapping(entity = Workspace.class, field = "members", parentIdField = "workspaceId")
public class WorkspaceMembers extends EntityList<String, Member> implements Workspace.Members {
  private String workspaceId;

  @Inject private WorkspaceMembersMapper mapper;
  @Inject private UsersMapper users;
  @Inject private WorkspacesMapper workspaces;
  @Inject private Clock clock;

  @Override
  protected List<Member> findEntities(int from, int to) {
    return mapper.findAll(workspaceId, from, Math.max(to - from, 0));
  }

  @Override
  protected Member findEntity(String id) {
    return mapper.findByIdentity(workspaceId, id);
  }

  @Override
  public int size() {
    return mapper.countAll(workspaceId);
  }

  @Override
  public Member add(MemberDescription description) {
    if (!workspaceId.equals(description.workspace().id())) {
      throw DomainException.validation(
          "member workspace "
              + description.workspace().id()
              + " does not match scoped workspace "
              + workspaceId);
    }
    String userId = description.user().id();
    if (users.findByIdentity(userId) == null) {
      throw DomainException.notFound("user " + userId + " not found");
    }
    if (workspaces.findByIdentity(workspaceId) == null) {
      throw DomainException.notFound("workspace " + workspaceId + " not found");
    }
    String memberId = UUID.randomUUID().toString();
    try {
      mapper.insert(
          memberId,
          workspaceId,
          new MemberDescription(
              description.workspace(),
              description.user(),
              WorkspaceAccess.role(description.role(), WorkspaceAccess.Role.MEMBER),
              description.createdAt(),
              description.updatedAt()),
          timestamp());
    } catch (DuplicateKeyException error) {
      throw DomainException.conflict("user " + userId + " is already a workspace member");
    }
    return require(memberId);
  }

  @Override
  public Member update(String memberId, String role) {
    Member current = require(memberId);
    String normalizedRole = WorkspaceAccess.role(role, null);
    assertOwnerRemains(current.getDescription().role(), normalizedRole);
    mapper.update(workspaceId, memberId, normalizedRole, timestamp());
    return require(memberId);
  }

  @Override
  public void remove(String memberId) {
    Member current = require(memberId);
    assertOwnerRemains(current.getDescription().role(), null);
    if (mapper.delete(workspaceId, memberId) != 1) {
      throw DomainException.notFound("workspace member " + memberId + " not found");
    }
  }

  private Member require(String memberId) {
    return findByIdentity(memberId)
        .orElseThrow(() -> DomainException.notFound("workspace member " + memberId + " not found"));
  }

  private void assertOwnerRemains(String currentRole, String nextRole) {
    if (!"owner".equals(currentRole) || "owner".equals(nextRole)) return;
    if (mapper.countOwners(workspaceId) <= 1) {
      throw DomainException.conflict("workspace must retain at least one owner");
    }
  }

  private Instant timestamp() {
    return clock.instant().truncatedTo(ChronoUnit.MILLIS);
  }
}
